### ISSUE TITLE
Wrong dualsense battery percentage

### ISSUE THREAD

**Issue by dkruuger** [Paliverse/DualSenseX#77]
**Describe the bug**
The battery percentage appears lower than it really is, while on DualSenseX it says 80%, on PS5 or DS4Windows it shows 100%

**Desktop:**
 - OS: [Windows]
 - Browser [chrome]
 - Version [21H2]

**Comment by Paliverse** [comment]
Yes, that has been known. It's hard to get an accurate reading without offical Drivers for the DualSense.

**Comment by Ivan-Alone** [comment]
It's shame that battery percentage is rather random than simple wrong. With cable connection shows 10% in DSX, in DS4Win - 12%; in wireless connection 40% and 50% respectively. 
While charging, can became to 40% for example, and drop again to 10% after reconnection USB.
Just bought DS today (not PS5), 'll leave it on charging to night, will check what happens.
I understand that we do not have documentation to this features, but battery level checker is extremly broken.
It's not hate of this awesome software, just bug report =)

**Comment by Paliverse** [comment]
The issue is the way data is being received from controller and the format we have to work with. It’s very bad. I really think it’s a bad design on Sony’s engineers

**Comment by DashCruft** [comment]
Maybe try making an artificial intelligence based on the data receiving on multiple devices that run the application? That could actually make accurate data maybe?

**Comment by Paliverse** [comment]
I understand what you mean, it's annoying not having an accurate reading, unfortunatly not much i can do about it, don't know how to make an AI either, out of my league lol

**Comment by KrisCris** [comment]
I think the battery percentage is capped at 80%, never seen the number go above it.

**Comment by DashCruft** [comment]
> I think the battery percentage is capped at 80%, never seen the number go above it.

I can confirm this

**Comment by Ivan-Alone** [comment]
> I think the battery percentage is capped at 80%, never seen the number go above it.
Is it matters in same situation? We have just only one trouble - gamepad tells wrong battery percentage at all, and what percentage is max found… Is really matters?
We all can confirm the same behavior. And it’s shame that those informations are useless.
Because Sony definitely lying in their Linux driver describing HID report in battery part.

**Comment by Ivan-Alone** [comment]
@Paliverse hmm, just seen that DS can tell truth, even about battery.
But question, why DSX shows wrong data? Need be fixed I think.
52 byte in HID report was 0x29, lets check official driver ([hid-playstation](https://aur.archlinux.org/cgit/aur.git/tree/hid-playstation.c?h=hid-playstation-dkms)):
```// ds_report->status is 52 byte (about charging)
battery_data = ds_report->status & DS_STATUS_BATTERY_CAPACITY; 
charging_status = (ds_report->status & DS_STATUS_CHARGING) >> DS_STATUS_CHARGING_SHIFT; 

switch (charging_status) {
...
case 0x2:
	battery_capacity = 100;
	battery_status = POWER_SUPPLY_STATUS_FULL;
```
So 0x29 must be interpreted like this: Battary charged, capacity is 100%. "9" in report must be ignored.
I checked your code via decompiller, and yes, you completely ignore ability of gamepad to say powerstate (discharging, charging, full charged or error). Only BT connection was handled.

And yes, this info does not changing fact that DS lies about battery status. But lets try to believe him)

**Comment by Paliverse** [comment]
@Ivan-Alone Decompiling my app is very rude to do so without my permission, you could have asked me how i read the data and i could have shown you. I went with what the linux driver had, as well as what another researcher that does some reverse engineering on the controller, told me the same info as well.

**Comment by Paliverse** [comment]
However i do see some differences between the driver and my code, will readjust and should be able get this sorted.

**Comment by DashCruft** [comment]
Great! - When can we expect this to be in DualSenseX? Also, when v1.5.0? 🤔

**Comment by Cphmi** [comment]
> Decompiling my app is very rude to do so without my permission, you could have asked me how i read the data and i could have shown you.

Tbh, it would be beneficial for such a nice project to make it open-sourced, there can be many people who willing to contribute.

**Comment by Paliverse** [comment]
@Cphmi while that could be nice, I’m not sure if I can work well with other devs, everyone has a different way of coding, so being able to understand what’s going on will become harder and harder as more contribute etc..  There’s also the design part of the app that I have done myself. Maybe one day it could potentially be open sourced if for whatever reason I’m unable to continue the project.

**Comment by Paliverse** [comment]
Just to retouch on this topic, DualSense Battery reading has been fixed for the next version releasing on Steam.

Make sure you also update the firmware from the official Firmwre Updater Tool released by sony: https://controller.dl.playstation.net/controller/lang/en/fwupdater.html
